image/manifest.go

```diff
@@ -26,28 +26,301 @@ import (
+// Have this here for use in unit tests.
+func NewSeedManifestForTest(revsAllowed, revsSeeded map[string]*SeedManifestSnapRevision, vsAllowed, vsSeeded map[string]*SeedManifestValidationSet) *SeedManifest {
+	sm := NewSeedManifest()
+	if revsAllowed != nil {
+		sm.revsAllowed = revsAllowed
+	}
+	if revsSeeded != nil {
+		sm.revsSeeded = revsSeeded
+	}
+	if vsAllowed != nil {
+		sm.vsAllowed = vsAllowed
+	}
+	if vsSeeded != nil {
+		sm.vsSeeded = vsSeeded
+	}
+	return sm
+}
+
+func (sm *SeedManifest) isControlledByValidationSet(snapName string) bool {
+	for _, vs := range sm.vsSeeded {
+		if vs.hasSnap(snapName) {
+			return true
+		}
+	}
+	return false
+}
+
+// SetAllowedSnapRevision adds a revision rule for the given snap name, meaning
+// that any snap marked used through MarkSnapRevisionUsed will be validated against
+// this rule. The manifest will only allow one revision per snap, meaning that any
+// subsequent calls to this will be ignored.
+func (sm *SeedManifest) SetAllowedSnapRevision(snapName string, revision int) error {
+	if revision == 0 {
+		return fmt.Errorf("cannot add a rule for a zero-value revision")
+	}
+
+	if _, ok := sm.revsAllowed[snapName]; !ok {
+		sm.revsAllowed[snapName] = &SeedManifestSnapRevision{
+			SnapName: snapName,
+			Revision: snap.R(revision),
+		}
+	}
+	return nil
+}
+
+// SetAllowedValidationSet adds a sequence rule for the given validation set, meaning
+// that any validation set marked for use through MarkValidationSetUsed must match the
+// given parameters. The manifest will only allow one sequence per validation set,
+// meaning that any subsequent calls to this will be ignored.
+func (sm *SeedManifest) SetAllowedValidationSet(accountID, name string, sequence int, pinned bool) error {
+	if sequence <= 0 {
+		return fmt.Errorf("cannot add allowed validation set %q for a unknown sequence",
+			fmt.Sprintf("%s/%s", accountID, name))
+	}
+
+	vs := &SeedManifestValidationSet{
+		AccountID: accountID,
+		Name:      name,
+		Sequence:  sequence,
+		Pinned:    pinned,
+	}
+	if _, ok := sm.vsAllowed[vs.Unique()]; !ok {
+		sm.vsAllowed[vs.Unique()] = vs
+	}
+	return nil
+}
+
+// MarkSnapRevisionSeeded attempts to mark a snap-revision as seeded in the manifest.
+// The seeded revision will be validated against any previously allowed revisions set. It
+// will also be validated against any revisions set in previously seeded validation sets.
+func (sm *SeedManifest) MarkSnapRevisionSeeded(snapName string, revision int) error {
+	snapRev := snap.R(revision)
+	if rev, ok := sm.revsAllowed[snapName]; ok {
+		// Allowed revision specified, it must match.
+		if rev.Revision != snapRev {
+			return fmt.Errorf("revision %s does not match the allowed revision %s", snapRev, rev.Revision)
+		}
+	}
+
+	if rev, ok := sm.revsSeeded[snapName]; ok {
+		// Already marked as seeding.
+		return fmt.Errorf("cannot mark %q (%s) as seeded, it has already been marked seeded: %q (%s)",
+			snapName, snapRev, rev.SnapName, rev.Revision)
```

Comment:
do we need to repeat the snap name twice?

Reply:
Ah nice catch, removed this!